yarn-project/noir-contracts/src/contracts/non_native_token_contract/src/main.nr

```diff
@@ -56,6 +70,176 @@ contract NonNativeToken {
+        recipient: Field,
+        _padding: [Field; dep::aztec3::abi::MAX_ARGS - 2]
+    ) {
+
+        let sender = inputs.call_context.msg_sender;
+        let sender_balance = public_balances.at(sender);
+
+        let current_sender_balance = sender_balance.read();
+
+        if (current_sender_balance as u120) > (amount as u120) {
+            let _void1 = sender_balance.write(current_sender_balance - amount);
+        }  
+        // TODO: Revert if there is not enough balance
+
+        let content = _get_withdraw_content_hash(amount, recipient);
+
+        // Emit the l2 to l1 message
+        create_l2_to_l1_message(content);
+    }
+
+
+    fn transfer(
+        //*********************************/
+        // Should eventually be hidden:
+        inputs: pub PrivateContextInputs,
+        //*********************************/
+        amount: pub Field, 
+        sender: pub Point, 
+        recipient: pub Point,
+    ) -> pub [Field; dep::aztec3::abi::PUBLIC_INPUTS_LENGTH] {
+        let mut initialContext = PrivateFunctionContext::new();
+        initialContext.args = initialContext.args.push_array([amount, sender.x, sender.y, recipient.x, recipient.y]);
+
+        let sender_balance = balances.at(sender.x);
+        let (mut context, (note1, note2)) = sender_balance.get_2(initialContext);
+
+        // TODO: Removing this return breaks the circuit (???)
+        context.return_values = context.return_values.push(note1.is_real as Field);
+        context.return_values = context.return_values.push(note2.is_real as Field);
+
+        note1.validate(sender);
+        note2.validate(sender);
+
+        let sum = note1.value + note2.value;
+        constrain sum as u64 >= amount as u64;
+        let change_value = sum - amount;
+
+        context = sender_balance.remove(inputs, context, note1);
+        context = sender_balance.remove(inputs, context, note2);
+
+        let change_note = Note::new(change_value, sender);
+        let recipient_note = Note::new(amount, recipient);
+
+        let recipient_balance = balances.at(recipient.x);
+
+        context = recipient_balance.insert(context, recipient_note);
+        context = sender_balance.insert(context, change_note);
+
+        constrain emit_encrypted_log(inputs.call_context.storage_contract_address, recipient_balance.storage_slot, recipient_note.owner, recipient_note) == 0;
+        constrain emit_encrypted_log(inputs.call_context.storage_contract_address, sender_balance.storage_slot, change_note.owner, change_note) == 0;
+
+        context.finish(inputs)
+    } 
+
+    /// ABI getBalance type "unconstrained"
+    fn getBalance(
+        owner: Point,
+        _padding: [Field; dep::aztec3::abi::MAX_ARGS - 2]
+    ) -> pub Field {
+        let owner_balance = balances.at(owner.x);
+        
+        balance_utils::get_balance(owner_balance.storage_slot)
+    }
+
+
+    fn _get_mint_content_hash(amount: Field, owner_address: Field, canceller: Field) -> pub Field {
```

Comment:
@kevaundray visibility for noir functions wen? 👀

Reply:
OkAy, OkAy, OkAy, I NeEd iNtErNaL FuNcTiOnS. i cAn't tAkE ThIs aNyMoRe. EvErYdAy i'm cHeCkInG ThE InTeRnAl fUnCtIoNs aNd iT'S NoT CoMpIlInG. eVeRyDaY I ChEcK ThE DoCs, BaD DoCs. I CaN'T TaKe tHiS AnYmOrE MaN. i hAvE OvEr-iNvEsTeD, bY A LoT. iT Is wHaT It iS BuT I NeEd tHe iNtErNaL FuNcTiOnS. cAn dEvS Do sOmEtHiNg?